### ISSUE TITLE
Is the Truenas option compatible with truenas Scale?

### ISSUE THREAD

**Issue by smooth-baking-powder** [jp-powers/fan-control#9]
Hi,

Is the Truenas option compatible with truenas Scale?

Thanks

**Comment by jp-powers** [comment]
Hi @corincorvus !

The TrueNAS option would be explicitly for TrueNAS Core. I **believe** TrueNAS Scale uses systemd, so it's quite possible the Proxmox install option would "just work" for it, but I have not tested that. I expect the Proxmox option would require far less adjusting to work on TrueNAS Scale to my understanding of it, but without a Scale box to do testing with myself I'm not comfortable saying anything with certainty.

Essentially, I don't know what is and is not included natively in TrueNAS Scale compared to Proxmox. Both Scale and Proxmox use Debian as their base, so there are certain things we can assume will be there, but iXSystems could have pulled out some things Proxmox leaves, or used alternatives. In other words, there are certain packages that may need to be installed to get the script to work, or could be located in difference places the script assumes are somewhere else, etc. I recall reading something from iXSystems while Scale was still in beta or RC that they _really_ suggested against installing or changing certain things in the underpinnings of the system, but I don't recall the full context around that.

At worst, if you try the Proxmox install option, it just won't work, and running the uninstall should clean it up. At best, it'll just work. Chances are it'll be somewhere in the middle, hopefully just small adjustments. If you want to give it a shot, let me know how it goes, and we can see if changes can be made.

-JP

**Comment by smooth-baking-powder** [comment]
Ok i will try. 

may you tell my, which packages are requiered? So i can proof first.

-CC

**Comment by jp-powers** [comment]
There's a few different things. Some of these will be there just due to what Scale is (it'd be weird if these aren't there), but it could be different versions or in a different location on the file system. That said, even if they are in different places or different versions, that should have very minor impact if any at all on how the script works.

TL;DR: The only package I'm not sure will be available in a base Scale install is lm-sensors, although I'd be surprised if they're using anything else, too. The only other concern I'd have is Python3 might be in a different path.

From install.sh, python needs to be in these locations depending on the platform because that's where it's found in a regular install for the platform:

```
TRUENASSHEBANG="#!/usr/local/bin/python3"
PROXMOXSHEBANG="#!/usr/bin/python3"
PFSENSESHEBANG="#!/usr/local/bin/python3.11"
```
You can run `whereis python3` to confirm where it is.

for Proxmox install, `/etc/systemd/system/` needs to be writable by root (should be anyways) but also not static. I believe iXSystems makes certain parts of the file system static, meaning they can't be changed, or at least effectively static, meaning after a reboot, update, or similar, they revert. I could maybe put in some research to figure this out but I'm short on time this morning, so if you're comfortable searching through the Scale docs I'm sure it'll be answered there. Alternatively, this can be somewhat tested by something like running the command `touch /etc/systemd/system/test.txt` and rebooting. 

In the actual script itself, it needs the following python modules available (most are standard, and if they're available in Core should be in Scale as well)

```
# import needed components
import time # used for sleep
import subprocess # used to execute external commands
import logging # lets make some nice logs
import json # we'll use json to handle reading the data from sensors a bit more cleanly/easily in Proxmox
import os # used to check if config.ini has changed
from configparser import ConfigParser # using ConfigParser for separate config file
import ast # used when importing config.ini to make it cleaner to work with
```
Lastly, the actual commands used in the script:
The `sensors` command is used to find CPU temps, and that comes from the `lm-sensors` package.
The `smartctl` command is used to find SATA drive times, and that comes from the `smartmontools` package.
The `ipmitool` command is used to set fan speeds, and that comes from the package of the same name, `ipmitool`.

**Comment by smooth-baking-powder** [comment]
Hey,

all your commands give me feedback, so i think, all is installed and ready. For ipmitool iam not sure.
```
root@s-1[~]# sensors
acpitz-acpi-0
Adapter: ACPI interface
temp1:        +27.8°C  (crit = +105.0°C)
temp2:        +29.8°C  (crit = +105.0°C)

coretemp-isa-0000
Adapter: ISA adapter
Package id 0:  +38.0°C  (high = +80.0°C, crit = +100.0°C)
Core 0:        +37.0°C  (high = +80.0°C, crit = +100.0°C)
Core 1:        +37.0°C  (high = +80.0°C, crit = +100.0°C)
Core 2:        +38.0°C  (high = +80.0°C, crit = +100.0°C)
Core 3:        +35.0°C  (high = +80.0°C, crit = +100.0°C)

root@s-1[~]# smartctl
smartctl 7.3 2022-02-28 r5338 [x86_64-linux-6.1.55-production+truenas] (local build)
Copyright (C) 2002-22, Bruce Allen, Christian Franke, www.smartmontools.org

ERROR: smartctl requires a device name as the final command-line argument.


Use smartctl -h to get a usage summary

root@s-1[~]# ipmitool
IANA PEN registry open failed: No such file or directory
Could not open device at /dev/ipmi0 or /dev/ipmi/0 or /dev/ipmidev/0: No such file or directory
```




`whereis python3`gives me:
```
root@s-1[~]# whereis python3
python3: 
/usr/bin/python3 
/usr/lib/python3 
/etc/python3 
/usr/share/python3 
/usr/share/man/man1/python3.1.gz

```
Its different to Proxmox.

I can test for you on my Truenas Scale Installation, if you want to add a new entry for truenas scale in your Menu.

**Comment by jp-powers** [comment]
I do see a problem in that output. impitool is showing no impitool devices found. Does your motherboard support ipmi?

**Comment by jp-powers** [comment]
Haven't heard back, assuming issue is effectively resolved. I'm potentially planning on changing up my TrueNAS setup in future with a new machine and likely will try TrueNAS Scale at the time, can test and make changes if needed then.
